Benchmark test case for q4_0 matrix multiplication

SebastianApel opened the pull request:
This is a small program I used benchmark the performance of the q4_0 matrix multiplication core.

I've used it to compare several implementations of `ggml_vec_dot_q4_0`.

I will reference this benchmark in another pull request for an improved AVX2 ggml_vec_dot_q4_0 implementation.

SebastianApel committed:
Initial version of q4_0 matrix multiplication benchmark

SebastianApel committed:
Merge branch 'ggerganov:master' into master

SebastianApel committed:
Bugfix: Added dependency to ggml.o to benchmark

howard0su commented on tests/test-benchmark-q4_0-matmult.c:
```diff
@@ -0,0 +1,212 @@
+        long long int start = rdtsc();
+        //printf("Running ggml_graph_compute\n");
+        ggml_graph_compute(ctx, &gf31);
+        long long int stop = rdtsc();
```
did you test it on windows? I don't think MSVC has rdtsc call. You can consider ggml_cycles and ggml_time_us

SebastianApel replied:
Switched to ggml_time_us

SebastianApel committed:
Reviewer requests: added parameter for threads, switched to ggml_time_us()

slaren commented on tests/test-benchmark-q4_0-matmult.c:
```diff
@@ -0,0 +1,258 @@
+#include <iterator>
+#include <algorithm>
+
+uint64_t rdtsc(){
```
This could probably be removed now that it is not used.

SebastianApel replied:
Done

slaren commented on tests/test-benchmark-q4_0-matmult.c:
```diff
@@ -0,0 +1,258 @@
+
+    // We cannot use the F32 result, because it will not be exactly the same (due to quantization)
+    // float sum_of_F32_reference = tensor_sum_elements(gf.nodes[0]);
+    float sum_of_F32_reference = 11611395072.00f;
```
Shouldn't we use an epsilon instead of hard-coding some value?

SebastianApel replied:
Done

slaren commented on tests/test-benchmark-q4_0-matmult.c:
```diff
@@ -0,0 +1,258 @@
+    printf("Iteration;NThreads; SizeX; SizeY; SizeZ; Required_FLOPS; Elapsed_u_Seconds; FLOPS_per_u_Second\n");
+    printf("==============================================================================================\n");
+    
+    setlocale(LC_ALL,"de_DE_UTF8");
```
Is this to ensure the decimal separator is what you expect? Could it be done in a more neutral way (instead of refering to de_DE)?

howard0su replied:
I don't think it is needed. You didn't use any non-ASCII characters. right?

SebastianApel replied:
Removed the setlocale

SebastianApel committed:
Reviewer input: removed rtsc, use epsilon for check

SebastianApel committed:
Review comment: Removed set_locale

slaren commented:
Looks good, but I am wondering if this should be split in two parts. On one hand we have a correctness test, which should be added to the list of tests to run in the CI with CTest, and on the other hand we have a benchmarking tool that is meant to be run manually, and should be somewhere else, maybe as one of the examples. What do you think?

SebastianApel committed:
Feature: Param for numer of iterations, Bugfix for use of parameter threads

SebastianApel committed:
Reviewer suggestion: Moved to examples

SebastianApel commented:
> Looks good, but I am wondering if this should be split in two parts. On one hand we have a correctness test, which should be added to the list of tests to run in the CI with CTest, and on the other hand we have a benchmarking tool that is meant to be run manually, and should be somewhere else, maybe as one of the examples. What do you think?

I've moved it to the folder "examples/benchmark" and adjusted the Makefile. Please note that I've not changed the CMakelists.

ggerganov commented on Makefile:
```diff
@@ -256,6 +256,10 @@ embedding: examples/embedding/embedding.cpp ggml.o llama.o common.o
 # Tests
 #
 
+benchmark: ggml.o
+	$(CXX) $(CXXFLAGS) tests/test-benchmark-q4_0-matmult.c ggml.o -o tests/test-benchmark-q4_0-matmult $(LDFLAGS)	
+	tests/test-benchmark-q4_0-matmult
+	
```
After moving to "examples" these seem to not have been updated.
Also update the `clean:` section

SebastianApel replied:
Done

ggerganov reviewed (changes requested):
It would be useful to have a mat mul benchmark tool, but I think we need more work to make this better:

- Add options to benchmark all data types, not just `Q4_0`
- Add lightweight correctness tests to CTest / CI
- I think no need for separate "example" and "test". A single "test" in the "./tests" folder that can be executed with different arguments should be best

I think this benchmark will eventually become part of the `ggml` repo.
For example, in `whisper.cpp` we already have similar benchmarks:

https://github.com/ggerganov/whisper.cpp/blob/master/whisper.cpp#L4686-L4774

So, if we add it here too, we will have duplication.

If we address the points above, we can merge the benchmark tool in this repo first and later move it to `ggml`.

SebastianApel committed:
Reviewer feedback: Updated clean: and benchmark: sections

SebastianApel commented:
@slaren suggested 
> maybe as one of the examples.

@ggerganov suggested
> I think no need for separate "example" and "test". A single "test" in the "./tests" folder that can be executed with different arguments should be best

Both options are fine with me - but I cannot fulfill both wishes. Who will decide and let me know which option should be implemented in order to get this PR merged?

SebastianApel commented:
@ggerganov 
> It would be useful to have a mat mul benchmark tool, but I think we need more work to make this better:
> - Add options to benchmark all data types, not just Q4_0
> - Add lightweight correctness tests to CTest / CI

I can understand the desire to extend the benchmark tool even more and enable it to work on all data types, and also to implement lightweight correctness tests to CTest / CI.

My challenge is that I will only have very limited capacity over the next weeks, so I myself will not be able to work on your suggestions.

I see the following options: 
- O1: We leave this PR open until someone finds the time to work on the suggestions
- O2: We merge the PR as is and open 1 or 2 "improvement" issue tickets for the suggestions
- O3: We close the PR and when someone finds the time to work on the suggestions, we will open a new one

Do you see any additional options? If not, which option would you prefer?

slaren commented:
@SebastianApel I am sorry for the confusion, I was mostly looking for input since I wasn't sure myself what was the best way to proceed. Absolutely follow @ggerganov suggestions, this is his project after all.

SebastianApel commented:
> @SebastianApel I am sorry for the confusion, I was mostly looking for input since I wasn't sure myself what was the best way to proceed. Absolutely follow @ggerganov suggestions, this is his project after all.

@slaren Thank you for the quick feedback. I can move it back, no problem. I'll wait for feedback regarding https://github.com/ggerganov/llama.cpp/pull/653#issuecomment-1493421622 before doing so.

thement commented:
> I see the following options:
> 
> * O1: We leave this PR open until someone finds the time to work on the suggestions
> * O2: We merge the PR as is and open 1 or 2 "improvement" issue tickets for the suggestions
> * O3: We close the PR and when someone finds the time to work on the suggestions, we will open a new one
> 
> Do you see any additional options? If not, which option would you prefer?

I would prefer O2.
I can't promise the time-frame but I would like to implement some correctness tests that would check the the optimized version against the reference version.

ggerganov committed:
Merge branch 'master' into master

jon-chuang commented:
Quite curious, perhaps add the `graph_print` after to show the individual node perf timings? (Requires GGML_PERF defined)

SebastianApel commented:
@jon-chuang Since this PR is merged - would you be open to  creating a new PR with your idea? That would make it easier to discuss your suggestion and/or code changes.
